akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala

```diff
@@ -0,0 +1,115 @@
+ * This back-off supervisor is created by using [[akka.pattern.BackoffSupervisor.props(options:akka\.pattern\.BackoffOptions)*]]
+ * with [[BackoffOptions.onFailure]].
+ */
+private class BackoffOnRestartSupervisor(
+  childProps: Props,
+  childName: String,
+  minBackoff: FiniteDuration,
+  maxBackoff: FiniteDuration,
+  reset: BackoffReset,
+  randomFactor: Double,
+  strategy: OneForOneStrategy)
+  extends Actor
+  with ActorLogging {
+
+  import BackoffOnRestartSupervisor._
+  import context._
+
+  override val supervisorStrategy = OneForOneStrategy(strategy.maxNrOfRetries, strategy.withinTimeRange, strategy.loggingEnabled) {
+    case ex ⇒
+      val defaultDirective: Directive =
+        super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) ⇒ Escalate)
+
+      strategy.decider.applyOrElse(ex, (_: Any) ⇒ defaultDirective) match {
+
+        // Whatever the final Directive is, we will translate all Restarts
+        // to our own Restarts, which involves stopping the child.
+        // directive match {
+        case Restart ⇒
+          val childRef = sender()
+          become({
+            case Terminated(`childRef`) ⇒
+              unbecome()
+              self ! ScheduleRestart(childRef)
+            case _ ⇒
+          }, discardOld = false)
+          Stop
+        case other ⇒ other
+      }
+  }
+
+  override def preStart(): Unit =
+    self ! StartChild
+
+  def receive = waitingToStart(-1, false)
+
+  def waitingToStart(numRestarts: Int, scheduleCounterReset: Boolean): Receive = {
+    case StartChild ⇒
+      val childRef = actorOf(childProps, childName)
+      watch(childRef)
+      if (scheduleCounterReset) {
+        system.scheduler.scheduleOnce(minBackoff, self, ResetRestartCount(numRestarts + 1))
+      }
+      become(watching(childRef, numRestarts + 1))
+    case msg ⇒ context.system.deadLetters.forward(msg)
+  }
+
+  // Steady state
+  def watching(childRef: ActorRef, numRestarts: Int): Receive = {
+    case ScheduleRestart(`childRef`) ⇒
+      val delay = BackoffSupervisor.calculateDelay(
+        numRestarts, minBackoff, maxBackoff, randomFactor)
+      system.scheduler.scheduleOnce(delay, self, StartChild)
+      become(waitingToStart(numRestarts, reset != ManualReset))
+      log.debug(s"Restarting child in: $delay; numRestarts: $numRestarts")
+
+    case ResetRestartCount(last) ⇒
+      if (last == numRestarts) {
+        log.debug(s"Last restart count [$last] matches current count; resetting")
+        become(watching(childRef, 0))
+      } else {
+        log.debug(s"Last restart count [$last] does not match the current count [$numRestarts]")
+      }
+
+    case BackoffSupervisor.Reset ⇒
+      reset match {
+        case ManualReset ⇒ become(watching(childRef, 0))
+        case msg         ⇒ unhandled(msg)
+      }
+    case BackoffSupervisor.GetCurrentChild ⇒
+      sender() ! BackoffSupervisor.CurrentChild(child(childName))
```

Comment:
Why not just: CurrentChild(Some(childRef))

- The child can always end up being shutdown after a response to GetCurrentChild.
- We should generally not be in the watching() state when the child is shutdown. See the sequence of lns 51 - 52, 82
- Which leads to another point, which is that you may want to handle the GetCurrentChild message in the waitingToStart() state and just respond with CurrentChild(None) ?